[ExtraShotLatte] iP

ExtraShotLatte opened the pull request:
### DukeNotSoPro
>“Your brain is your mind.” – Some random guy on the internet ([source](https://google.com))

DukeNotSoPro makes you try to remember things you have not done.

_It's:_
1. text-based
2. easy to **learn**
3. **FAST** _SUPER FAST_ to use
4. All ~you~ need to do is,


download it from [here](https://nus-cs2103-ay2223s2.github.io/website/schedule/week4/project.html).
- double-click it.
- add your tasks.
- let it manage your tasks for you 😉
- And it is FREE!

Features:
- [ ] Managing tasks
- [x]  Managing deadlines (coming soon)
- [x]  Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

ExtraShotLatte committed:
Create initial prompt

ExtraShotLatte committed:
Fix spacing for initial prompt

ExtraShotLatte committed:
Add scanner structure

ExtraShotLatte committed:
implement echo-ing of user commands

ExtraShotLatte committed:
implement bye message

ExtraShotLatte committed:
change variable of input for easier understanding

ExtraShotLatte committed:
Add switch + loop to print tasks

ExtraShotLatte committed:
clean up code for level 2 submission

ExtraShotLatte committed:
Create basic task implementation

ExtraShotLatte committed:
Refactor code to use task object instead of arraylist

ExtraShotLatte committed:
Refactor code to better manage keywords

ExtraShotLatte committed:
Change marker for done task

ExtraShotLatte committed:
Add mark command

ExtraShotLatte committed:
Add unmark command

ExtraShotLatte committed:
Fix list to start from index 1

ExtraShotLatte committed:
Clean up code

ExtraShotLatte committed:
Add ToDo class

ExtraShotLatte committed:
Refactor: Use ArrayList instead of array

ExtraShotLatte committed:
Fix typo in variable name

ExtraShotLatte committed:
Add tracking of todo

ExtraShotLatte committed:
Add Deadline class

ExtraShotLatte committed:
Implement task type: deadline

ExtraShotLatte committed:
Add event class

ExtraShotLatte committed:
Implement Task: Event

ExtraShotLatte committed:
Semi-automate the testing of Duke

ExtraShotLatte committed:
Add Duke Exception Class

ExtraShotLatte committed:
Add exception handling for empty descriptions

ExtraShotLatte committed:
Clean up code

ExtraShotLatte committed:
Add support for delete function

ExtraShotLatte committed:
Add new storage class

ExtraShotLatte committed:
Add Storage test

ExtraShotLatte committed:
Add skeletal load method for Storage

ExtraShotLatte committed:
Add TaskList class

ExtraShotLatte committed:
Add printTask method to TaskList class

ExtraShotLatte committed:
Add support for deletion of task in TaskList

ExtraShotLatte committed:
Add support for printNewestTask to TaskList

ExtraShotLatte committed:
Add mark, unmark and delete task to TaskList

ExtraShotLatte committed:
Refactor main code to use TaskList

ExtraShotLatte committed:
Add asCSV method for Event

ExtraShotLatte committed:
Add toCSV method for ToDo

ExtraShotLatte committed:
Add toCSV method for Deadline

ExtraShotLatte committed:
Fix deadline asCSV method

ExtraShotLatte committed:
Add loading of task based on CSV

ExtraShotLatte committed:
Add boolean to constructor for Task

ExtraShotLatte committed:
Update Deadline to allow support of intialising Boolean

ExtraShotLatte committed:
Update of ToDo to support initialising of boolean

ExtraShotLatte committed:
Update of Event to support initialising of boolean

ExtraShotLatte committed:
Update tasklist to support boolean constructor

ExtraShotLatte committed:
Update storage to support loading of boolean

ExtraShotLatte committed:
Add CSV Abstract method to Task

ExtraShotLatte committed:
Implement saving of task

ExtraShotLatte committed:
Implement loading and saving of task

ExtraShotLatte committed:
Add saving of task to mark and unmark command

ExtraShotLatte committed:
Add datetime to Deadline

ExtraShotLatte committed:
Merge branch 'branch-Level-7'

ExtraShotLatte committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java

ExtraShotLatte committed:
Clean up code in storage

ExtraShotLatte committed:
Add Ui Class

ExtraShotLatte committed:
Add UI class

ExtraShotLatte committed:
Allow storage to throw exception on loading error

ExtraShotLatte committed:
Add Parser class

ExtraShotLatte committed:
Add read command to UI

ExtraShotLatte committed:
Add exit and scanner close methods to UI

ExtraShotLatte committed:
Allow parser to parse different commands

ExtraShotLatte committed:
Refactor Duke to use parser and UI elements for commands

ExtraShotLatte committed:
Refactor duke into packages

ExtraShotLatte committed:
Merge branch 'add-gradle-support' of https://github.com/ExtraShotLatte/ip

ExtraShotLatte committed:
Clean up imports of Storage

ExtraShotLatte committed:
Add comons-io to Gradle Dependencies

ExtraShotLatte committed:
Add JUnit test for events

ExtraShotLatte committed:
Add JUnit test for ToDo

ExtraShotLatte committed:
Add manifest file to gitignore

ExtraShotLatte committed:
Clean up main code

ExtraShotLatte committed:
Remove uneeded comment in Duke

ExtraShotLatte committed:
Add java docs for main Duke class

ExtraShotLatte committed:
Add Javadocs for Task class

ExtraShotLatte committed:
Add Javadocs for Duke main class

ExtraShotLatte committed:
Add JavaDocs to Deadline

ExtraShotLatte committed:
Add JavaDocs for ToDo

ExtraShotLatte committed:
Add JavaDocs for TaskList

ExtraShotLatte committed:
Add JavaDocs for Storage

ExtraShotLatte committed:
Add JavaDocs for Parser

ExtraShotLatte committed:
Add JavaDocs for Event

ExtraShotLatte committed:
Add JavaDocs for DukeException

ExtraShotLatte committed:
Refactor Duke to use course naming conventions

ExtraShotLatte committed:
Refactor ToDo to match naming convention

ExtraShotLatte committed:
Refactor Parser to match naming convention

ExtraShotLatte committed:
Refactor Duke Exception to match naming convention

ExtraShotLatte committed:
Refactor Event to match naming convention

ExtraShotLatte committed:
Refactor Duke to match naming convention

ExtraShotLatte committed:
Refactor Deadline to match naming convention

ExtraShotLatte committed:
Change task name to package protected

ExtraShotLatte committed:
Implement search feature

ExtraShotLatte committed:
Merge branch 'A-JavaDoc'

ExtraShotLatte committed:
Merge branch 'A-CodingStandard'

# Conflicts:
#	src/main/java/Duke.java
#	src/main/java/duke/DukeException.java
#	src/main/java/duke/Parser.java
#	src/main/java/duke/ToDo.java

ExtraShotLatte committed:
Merge branch 'Level-9'

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Task.java

ExtraShotLatte committed:
Add checkstyle support

ExtraShotLatte committed:
Add JavaDocs to Ui

ExtraShotLatte committed:
Add JavaDocs and fix checkstyle

ExtraShotLatte committed:
Merge branch 'A-CheckStyle'

leonpoh commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,65 @@
+/**
+ * Models an deadline which is a task.
+ */
+public class Deadline extends Task {
```
Looks good seperating task into its subclasses

leonpoh reviewed (commented):
I like that you abstracted out the Task, however, maybe you should create another class that executes those command.

ExtraShotLatte committed:
Add initial files for JavaFX

shenchenzizoe commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,66 @@
+
+        new Duke("tasks.txt").runDuke();
     }
 }
```
nice OOP implemented

shenchenzizoe commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,65 @@
+        super(desc, b);
+        this.desc = desc;
+        this.by = by;
+        byDate = LocalDate.parse(by);
```
could consider handle exceptions when String by does not meet the LocalDate format?

ryanchua00 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,122 @@
+            storage.save(tasks);
+            break;
+        case "mark":
+            selectedNum = Integer.parseInt(inputCmdArr[1]);
+            tasks.markTask(selectedNum);
+            storage.save(tasks);
+            break;
+        case "unmark":
+            selectedNum = Integer.parseInt(inputCmdArr[1]);
```
Would this throw a NumberFormatException?

shenchenzizoe commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,65 @@
+     */
+    @Override
+    public String toString() {
+        return "[D]" + super.toString() + " (by: " + byDate.format(DateTimeFormatter.ofPattern("MMM d yyyy")) + ")";
```
splitting into two lines might meet the coding standards better

ryanchua00 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,110 @@
+        if (Files.exists(filePath)) {
+            System.out.println("FILE EXIST");
+            TaskList loadTaskList= new TaskList();
+            try {
+                List<String> allLines = Files.readAllLines(filePath);
+                for (String line : allLines) {
+                    String[] lineArray = line.split(",");
+                    boolean b;
+                    if (lineArray[1].equals("1")) {
+                        b = true;
+                    } else {
+                        b = false;
+                    }
+                    switch (lineArray[0]) {
+                        case "T":
+                            loadTaskList.addTask(lineArray[2], b);
+                            break;
+                        case "D":
+                            loadTaskList.addTask(lineArray[2], lineArray[3], b);
+                            break;
+                        case "E":
+                            loadTaskList.addTask(lineArray[2], lineArray[3], lineArray[4], b);
+                            break;
+                        case "":
+                            break;
+                    }
+                }
+                return loadTaskList;
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        } else {
+            System.out.println("NO EXIST");
+            try {
+                FileUtils.write(dukeDataFile, "");
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
```
Can this try-catch block be moved outside of the if-else?

ryanchua00 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,67 @@
+     *
+     * @return Task in CSV representation.
+     */
+    public abstract String asCsv();
```
I like this! 
Although it does necessitate storing "desc" as a field in every child class.

shenchenzizoe commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,67 @@
+    /**
+     * Method to unmark task as done.
+     */
+    public void markUndone() {
```
may fit the required command better if change method name to "mark & unmark"

ryanchua00 reviewed (commented):
Overall, great code. It was easy to read and understand. I also understand that with the iP we code more to make it work and have little to no time to go back to improve things. So with the things I picked out, they are just small optimizations for the most part.

shenchenzizoe reviewed (commented):
Generally the appllication of OOP, the use of packages, and the use of Enums makes the code structured and quite readable! Just some minor coding standard issues to notice.

ExtraShotLatte committed:
Refactor to make UI into subpackage

ExtraShotLatte committed:
Add Mainwindow class controller

ExtraShotLatte committed:
Add DialogBox class

ExtraShotLatte committed:
Add Main Class for setting up JavaFX Stage

ExtraShotLatte committed:
Add Launcher as point of entry to draw GUI

ExtraShotLatte committed:
Add methods to return result of methods in TaskList as String

ExtraShotLatte committed:
Add new function in parser that will return the output of command as String

ExtraShotLatte committed:
Remove dummy variable in mainwindow to enable it to use Duke

ExtraShotLatte committed:
Add method in duke that returns it's response in String

ExtraShotLatte committed:
Merge branch 'Level-10'

ExtraShotLatte committed:
Update Build.Gradle file to include JavaFX dependencies

This is needed to generate a far JAR using the ShadowJar command of
gradle. Additionally, we may not assume that users who downloads our
jar file have the correct dependencies already installed.

We need to,
* ensure that our JAR can cater to most if not everyone
* ensure that everyone get the same experience on different platform

ExtraShotLatte committed:
Enable assertions in Gradle

ExtraShotLatte committed:
Assert that duke always have a response

ExtraShotLatte committed:
Assert parser will always return a result

ExtraShotLatte committed:
Remove unused code after JavaFX implementation

ExtraShotLatte committed:
Add comment

ExtraShotLatte committed:
Refactor to use better variable name in Deadline class

ExtraShotLatte committed:
Refactor to use better variable name in Event class

ExtraShotLatte committed:
Refactor to remove unnecessary methods after shifting to JavaFX

ExtraShotLatte committed:
Refactor Storage to seperate methods for each abstraction

ExtraShotLatte committed:
Convert class specific methods to private access modifier

ExtraShotLatte committed:
Add JavaDocs for newly abstracted methods from Storage

ExtraShotLatte committed:
Refactor variable to better reflect status of task

ExtraShotLatte committed:
Refactor task to use better variable names

ExtraShotLatte committed:
Delete unused obselete methods in TaskList

ExtraShotLatte committed:
Refactor main input window to not accept blank / empty command

ExtraShotLatte committed:
Add abstract Command class

ExtraShotLatte committed:
Add DeadlineCommand class

ExtraShotLatte committed:
Add DeleteCommand class

ExtraShotLatte committed:
Add EventCommand class

ExtraShotLatte committed:
Add ListCommand class

ExtraShotLatte committed:
Add MarkCommand class

ExtraShotLatte committed:
Add UnmarkCommand class

ExtraShotLatte committed:
Add ToDoCommand class

ExtraShotLatte committed:
Add SearchCommand class

ExtraShotLatte committed:
Refactor parser to abstract commands

ExtraShotLatte committed:
Fix commands to properly cache index to be operated on

ExtraShotLatte committed:
Refactor commands to better handle and throw exceptions

ExtraShotLatte committed:
Refactor code to better handle and throw exceptions

ExtraShotLatte committed:
Change picture of user and bot

ExtraShotLatte committed:
Merge pull request #2 from ExtraShotLatte/branch-A-Assertions

Branch a assertions

ExtraShotLatte committed:
Merge branch 'master' of https://github.com/ExtraShotLatte/ip into branch-A-CodeQuality

# Conflicts:
#	src/main/java/Duke.java

ExtraShotLatte committed:
Merge pull request #3 from ExtraShotLatte/branch-A-CodeQuality

Branch-A-CodeQuality

ExtraShotLatte committed:
Integrate CI

ExtraShotLatte committed:
Organize Parser and remove unused methods

ExtraShotLatte committed:
Add JavaDoc for main Parser method

ExtraShotLatte committed:
Remove unused UI parameters

ExtraShotLatte committed:
Refactor command package to use small caps

ExtraShotLatte committed:
Fix style of code

ExtraShotLatte committed:
Add comment for C-BetterSearch

ExtraShotLatte committed:
Add Ui screenshot

ExtraShotLatte committed:
Add title for main window

ExtraShotLatte committed:
Fix released JAR not running properly

ExtraShotLatte committed:
Change userguide

ExtraShotLatte committed:
Fix attempt: rename UI png

ExtraShotLatte committed:
Fix attempt - to use Ui.png (small letter)

ExtraShotLatte committed:
Remove text-UI test

ExtraShotLatte committed:
Update CI workflow to not perform Text-UI test

ExtraShotLatte committed:
Update UG to include proper links to JAR and JDK

ExtraShotLatte committed:
Update Gradle to use JavaFX Version 11.0.2
